x/gamm/pool-models/balancer/amm_test.go

```diff
@@ -436,3 +436,683 @@ func TestCalcJoinPoolShares(t *testing.T) {
+		})
+	}
+}
+
+func TestCalcJoinSingleAssetTokensIn(t *testing.T) {
+	testCases := []struct {
+		name           string
+		swapFee        sdk.Dec
+		poolAssets     []balancer.PoolAsset
+		tokensIn       sdk.Coins
+		expectShares   sdk.Int
+		expectLiqudity sdk.Coins
+	}{
+		{
+			// Expected output from Balancer paper (https://balancer.fi/whitepaper.pdf) using equation (25) on page 10:
+			// P_issued = P_supply * ((1 + (A_t / B_t))^W_t - 1)
+			//
+			// 2_499_999_968_750 = 100 * 10^18 * (( 1 + (50,000 / 1_000_000_000_000))^0.5 - 1)
+			//
+			// where:
+			// 	P_supply = initial pool supply = 100 * 10^18 (set at pool creation, same for all new pools)
+			//	A_t = amount of deposited asset = 50,000
+			//	B_t = existing balance of deposited asset in the pool prior to deposit = 1,000,000,000,000
+			//	W_t = normalized weight of deposited asset in pool = 0.5 (equally weighted two-asset pool)
+			// Plugging all of this in, we get:
+			// 	Full solution: https://www.wolframalpha.com/input?i=100000000000000000000*%28%281+%2B+%2850000%2F1000000000000%29%29%5E0.5+-+1%29
+			// 	Simplified:  P_issued = 2,499,999,968,750
+			name:    "one token in - equal weights with zero swap fee",
+			swapFee: sdk.MustNewDecFromStr("0"),
+			poolAssets: []balancer.PoolAsset{
+				{
+					Token:  sdk.NewInt64Coin("uosmo", 1_000_000_000_000),
+					Weight: sdk.NewInt(100),
+				},
+				{
+					Token:  sdk.NewInt64Coin("uatom", 1_000_000_000_000),
+					Weight: sdk.NewInt(100),
+				},
+			},
+			tokensIn:     sdk.NewCoins(sdk.NewInt64Coin("uosmo", 50_000)),
+			expectShares: sdk.NewInt(2_499_999_968_750),
+		},
+		{
+			// Expected output from Balancer paper (https://balancer.fi/whitepaper.pdf) using equation (25) on page 10:
+			// P_issued = P_supply * ((1 + (A_t / B_t))^W_t - 1)
+			//
+			// 2_499_999_968_750 = 100 * 10^18 * (( 1 + (50,000 / 1_000_000_000_000))^0.5 - 1)
+			//
+			// where:
+			// 	P_supply = initial pool supply = 100 * 10^18 (set at pool creation, same for all new pools)
+			//	A_t = amount of deposited asset = 50,000
+			//	B_t = existing balance of deposited asset in the pool prior to deposit = 1,000,000,000,000
+			//	W_t = normalized weight of deposited asset in pool = 0.5 (equally weighted two-asset pool)
+			// Plugging all of this in, we get:
+			// 	Full solution: https://www.wolframalpha.com/input?i=100000000000000000000*%28%281+%2B+%2850000%2F1000000000000%29%29%5E0.5+-+1%29
+			// 	Simplified:  P_issued = 2,499,999,968,750
+			name:    "two tokens in - equal weights with zero swap fee",
+			swapFee: sdk.MustNewDecFromStr("0"),
+			poolAssets: []balancer.PoolAsset{
+				{
+					Token:  sdk.NewInt64Coin("uosmo", 1_000_000_000_000),
+					Weight: sdk.NewInt(100),
+				},
+				{
+					Token:  sdk.NewInt64Coin("uatom", 1_000_000_000_000),
+					Weight: sdk.NewInt(100),
+				},
+			},
+			tokensIn:     sdk.NewCoins(sdk.NewInt64Coin("uosmo", 50_000), sdk.NewInt64Coin("uatom", 50_000)),
+			expectShares: sdk.NewInt(2_499_999_968_750 * 2),
```

Comment:
Are you sure you can multiply (* 2) to calculate shares for two asset deposit with same amount of tokens In? 

cc @AlpinYukseloglu @alexanderbez

Reply:
> Are you sure you can add both shares to calculate total shares for two asset deposit ?
> 
> cc @AlpinYukseloglu @alexanderbez

Reply:
As per my comment above, https://github.com/osmosis-labs/osmosis/pull/1721/files#r893463274

here, we can add shares because these are essentially done by single asset joining each token.

Reply:
Seems reasonable to me.

Reply:
I'm confused as to why `expectShares` is double what is in the previous test when the total liquidity isn't being doubled. From the [balancer whitepaper](https://balancer.fi/whitepaper.pdf):

> An “all-asset” deposit has to follow the distribution of existing assets in the pool. If the deposit contains 10%
of each of the assets already in the pool, then the Value Function will increase by 10% and the depositor will
be minted 10% of the current outstanding pool token supply.

Also as per my comment above, it's not clear to me why we're implementing/testing a case where we single asset join two assets into a 2-asset pool when `MaximalExactRatioJoin` would do the trick with no fees (and would be the intended/expected behavior when a pool is joined)

Reply:
> I'm confused as to why expectShares is double what is in the previous test when the total liquidity isn't being doubled

The logic was the following - here we are adding 2 tokens in that have equal weights and amounts in, so the liquidity would double. In the previous case, there was only one token in.

However, it seems that these should still not be equal because after the first join we would have more shares created for one of the tokens, right? As a result, the second join would have fewer shares returned than the first one. Seems that this test is passing because of the error tolerance

Reply:
> Also as per my comment above, it's not clear to me why we're implementing/testing a case where we single asset join two assets into a 2-asset pool when MaximalExactRatioJoin would do the trick with no swap fees (and would be the intended/expected behavior when a pool is joined) – am I missing something

The goal was to make the function be independent of what the caller might do. I think the suggested check is useful but should belong to the caller of this function. Let me know what you think

Reply:
> > I'm confused as to why expectShares is double what is in the previous test when the total liquidity isn't being doubled
> 
> The logic was the following - here we are adding 2 tokens in that have equal weights and amounts in, so the liquidity would double. In the previous case, there was only one token in.
> 
> However, it seems that these should still not be equal because after the first join we would have more shares created for one of the tokens, right? As a result, the second join would have fewer shares returned than the first one. Seems that this test is passing because of the error tolerance

This is interesting – I wonder if we can set up a case where this falls outside of error tolerance just for proper coverage? An example could be deposit amounts that are very large relative to the pool liquidity (i.e. in this case we're depositing 50_000 uosmo into a pool with 1_000_000_000_000 uosmo – what if the amount we deposit was much higher so the 2nd transaction generates significantly fewer LP shares than the first?)

Reply:
> > Also as per my comment above, it's not clear to me why we're implementing/testing a case where we single asset join two assets into a 2-asset pool when MaximalExactRatioJoin would do the trick with no swap fees (and would be the intended/expected behavior when a pool is joined) – am I missing something
> 
> The goal was to make the function be independent of what the caller might do. I think the suggested check is useful but should belong to the caller of this function. Let me know what you think

I hear you on this, agreed!